pkg/hintrunner/hint.go

```diff
@@ -88,3 +90,78 @@ func (hint TestLessThan) Execute(vm *VM.VirtualMachine) error {
 
 	return nil
 }
+
+type WideMul128 struct {
+	lhs  ResOperander
+	rhs  ResOperander
+	high CellRefer
+	low  CellRefer
+}
+
+func (hint WideMul128) String() string {
+	return "WideMul128"
+}
+
+func (hint WideMul128) Execute(vm *VM.VirtualMachine) error {
+	mask := MaxU128Felt()
+
+	lhs, err := hint.lhs.Resolve(vm)
+	if err != nil {
+		return fmt.Errorf("resolve lhs operand %s: %v", hint.lhs, err)
+	}
+
+	rhs, err := hint.rhs.Resolve(vm)
+	if err != nil {
+		return fmt.Errorf("resolve rhs operand %s: %v", hint.rhs, err)
+	}
+
+	lhsFelt, err := lhs.FieldElement()
+	if err != nil {
+		return err
+	}
+	if lhsFelt.Cmp(&mask) > 0 {
+		return fmt.Errorf("lhs operand %s should be u128", lhsFelt)
+	}
+
+	rhsFelt, err := rhs.FieldElement()
+	if err != nil {
+		return err
+	}
+	if rhsFelt.Cmp(&mask) > 0 {
+		return fmt.Errorf("rhs operand %s should be u128", rhsFelt)
+	}
+
+	lhsU256 := uint256.Int(lhsFelt.Bits())
+	rhsU256 := uint256.Int(rhsFelt.Bits())
+	mul := lhsU256.Mul(&lhsU256, &rhsU256)
+
+	bytes := mul.Bytes32()
+
+	low := f.One()
+	low.SetBytes(bytes[16:])
+
+	high := f.One()
+	high.SetBytes(bytes[:16])
```

Comment:
are we sure low and high parts are equal size (ie 16 bytes)? and not low being all the lower 251 bit and high being the rest 5 most significant bits.

Reply:
I'm not sure what you mean here, we multiply two u128 together, resulting in a value that goes up to 2^256 - 2*2^128 + 1. We then split this value into low part (so 16 last bytes in little-endian) and the high part (so 16 first bytes in little-endian). I used the Bytes32 function from u256 to be sure the returned array is of size 32.